SVA/lib/handlers.S

```diff
@@ -507,24 +507,58 @@ SVAsyscall:
    * if this is the fork(), vfork(), rfork(), or pdfork() system calls.  The
    * system call number will be in %rax and can have the following values
    * (you can find these in syscalls.master) within the FreeBSD kernel source
-   * code):
+   * code).
+   * According to the type of fork and the flags, we make a decision about
+   * whether to COW on ghost memory or not, and save the decision in the flags
+   * filed of the current SVAThread. (1 is COW, and 0 is not COW.)
+   * In particular, fork() and pdfork() always incur COW, vfork() does not incur COW,
+   * and for rfork() it depends on its arguments.
    *
    *   2 - fork()
    *  66 - vfork()
    * 251 - rfork()
    * 518 - pdfork()
    */
+
+  /* Find the flags field of the current SVAThread */
+  movq %gs:0x260, %r14          
+  movq CPU_THREAD(%r14), %r14    
+  movq $0,  0x7d80(%r14)  
+  movq $0, %rbx
+  movq $1, %r11
+
   movq $1, %r12
   movq $3, %r13
   cmpq $2, IC_RAX(%rsp)
   cmoveq %r13, %r12
+  cmoveq %r11, %rbx
+  je DONE_FORK
+
   cmpq $66, IC_RAX(%rsp)
   cmoveq %r13, %r12
+  je DONE_FORK
+
   cmpq $251, IC_RAX(%rsp)
-  cmoveq %r13, %r12
+  jne NEXT_TEST_FORK
+  movq %r13, %r12
+  /* test the syscall argument flags, whether 
```

Comment:
It looks like you're testing the flags argument to see if the ghost memory should be copy-on-write.  Is that correct?  If so, we're adding overhead to the critical path (system call dispatch).  The flags argument of the system call is already saved in the Interrupt Context; we could have sva_init_stack() just check the flags there to determine if the ghost memory should be marked copy-on-write.  Adding overhead to sva_init_stack would be better than adding overhead to system call dispatch.